### PR TITLE
Improve PdfSharpTextExtractor Initial Checkin

### PR DESCRIPTION
**Pull request by cleversolutions**
* Character substitution using Encoding Differences array
 * Character substitution using CMaps

I wanted to get a draft PR in so I can start sharing this work with others. Reviewing the diffs it looks like I need to at least add some more comments. I'm sure there are some places where exceptions can get tossed as well which I'll have to clean up. I'm currently testing these changes on one of our new sites that is still under development. I should have a better idea how solid this is after indexing 4800+ PDFs on it. 

Resolves issue #18 

---
_This item has been added to our backlog [AB#9469](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/9469)_

### CODE REVIEW

**Comment by Shazwazza** [comment]
any update on the testing? not sure where you found this stuff, looks interesting :)

**Comment by cleversolutions** [comment]
It is interesting stuff, I had a lot of fun building it. It is all based on the PDF specs. I'm certainly not an expert on PDF, but I now understand it a lot better than I had a year ago. It is certainly better than it was, but still not perfect. There are a few cases where it doesn't catch newlines so words run together, but at least it is now extracting those words where previously it would not have. I would rather merge this and have the better functionality and address any smaller problems in new issues.

**Comment by cleversolutions** [comment]
I had been testing this on a couple of sites, but my package got replaced at some point by with the official package. I've bumped the version up to 2.1 and re-installed it into our sites for further testing.

**Comment by cleversolutions** [comment]
It logs an informational message for each pdf it indexes. Do you think this is going to get annoying?

**Comment by cleversolutions** [comment]
Wait never mind, it doesn't, that was a custom component of mine that does.

**Comment by cleversolutions** [comment]
I added some comments to the code. I wish I had done this a couple of months ago while it was all fresh in my mind.

**Comment by cleversolutions** [comment]
I've re-indexed over 10k PDFs on https://www.northbay.ca and had only 2 exceptions thrown which are caused by the PDF documents being protected. We can't get anything out of a protected document anyway, so no big loss.

**Comment by Shazwazza** [comment]
Sorry for the delay again! But that is just absolutely amazing news and this work is quite incredible 🎉
In this case it really does seem like all that's needed is maybe a bit of code cleanup and review and get this thing shipped, what do you think?

**Comment by cleversolutions** [comment]
Thanks @Shazwazza. I'll try to tidy the code up a bit this week.

**Comment by cleversolutions** [comment]
@Shazwazza what are your thoughts on the AdobeGlyfList.cs? I embedded the Adobe Glyph List file as a giant string in the CS file so I didn't have to deal with external files. Should we leave it as is or read it as a file and store that file somewhere (bin directory?)? I think if memory serves me correctly, I was struggling to get Nuget to install the file in the bin, so I embeded it as a string in the cs file.

**Comment by cleversolutions** [comment]
I figured out how to put the Adobe Glyph List file in the App_Plugins folder.

**Comment by Shazwazza** [comment]
IMO it would just be simpler to have that as an embedded resource. Perhaps you could then have an option to configure an external file if people didn't want to use the default?

**Comment by cleversolutions** [comment]
That sounds like a good idea. I'll leverage IOC to make this much more configurable.

**Comment by cleversolutions** [comment]
@Shazwazza I've leveraged IOC to make this more configurable and inject services into the various pieces. I've replaced the Debug.WriteLine calls with logger.debug calls. Clean up a few things and added some more comments. Take a look and let me know what you think.

**Comment by NijasHameed** [comment]
@cleversolutions 
I have tried your solution to extract text from pdf but I am getting some small errors like Length cannot be less than zero etc., on CMAP Character substitution, Can you try to extract text with the attached pdf file and get a solution.[
[file-example_pdf_1mb.pdf](https://github.com/umbraco/UmbracoExamine.PDF/files/5190468/file-example_pdf_1mb.pdf)
](url)

**Comment by NijasHameed** [comment]
And also Some PDFs like the attached is not encoding. 
[file-example_pdf_1mb.pdf](https://github.com/umbraco/UmbracoExamine.PDF/files/5190547/file-example_pdf_1mb.pdf)

**Comment by cleversolutions** [comment]
Thanks @NijasHameed I will run some tests with these files. There are a lot of known limitations with this text processor at the moment, I hope to improve it over time, but I doubt it will ever be as good as iText. If you need a better text processor, and AGPL isn't a problem for your use case, you can plug in iText easily enough.

**Comment by NijasHameed** [comment]
Great 👍 Thanks @cleversolutions , I will try with iText.

**Comment by cleversolutions** [comment]
file-example_pdf_1mb.pdf should now be working, that found a typo in my code. Oops. 11.pdf and 12.pdf are exposing some other more interesting problems. I'm not sure 12.pdf will ever work. Adobe Reader can't read text out of it so how could a 3rd party plugin, but 11.pdf should be readable. In the case of 11.pdf the text is perfect, then I encode it through the differences array and it becomes a disaster.  Perhaps I'm missing some mechanism to determine when I should and shouldn't be using the differences array to encode characters. Sigh, PDF is hard.

@Shazwazza I'm going to look into a different direction. PdfPig shows a lot of promise in doing text extraction out of the box. I'm going to investigate using it instead of PdfSharp. The text extraction could would be about a handful of lines.

**Comment by Shazwazza** [comment]
@cleversolutions awesome work! let us know how the PdfPig integration goes. I mean the current state is WAY better than before, i'd be tempted to just ship what you've done soon but will hold off for now. ... but what do you think?

**Comment by cleversolutions** [comment]
@Shazwazza this has definitely evolved. This PR now drops PdfSharp for PdfPig and amounts now to just a handful of lines changed. All unit tests passed. I'm deploying this across some production sites for further testing.

**Comment by Shazwazza** [comment]
@cleversolutions sounds cool, let me know how you go, happy to start looking at getting a release out once you give me the green light to test :)

**Comment by cleversolutions** [comment]
Hi @Shazwazza I think this is ready to go. I'm using it on our new website which goes live on Thursday. Specifically on this page https://www.northbay.ca/city-government/by-laws/search-browse/ we are indexing about 7500 bylaws since 1913 with no errors. I've spot checked a number of PDFs, the only ones that haven't indexed are scanned images that were not OCRed. The OCRed bylaws process just fine, so do the files converted from word. Documents we received from vendors work great, as do ones converted from InDesign.

**Comment by Shazwazza** [comment]
Amazing! I have this starred in my inbox and will try to get to this this week. Awesome stuff :)

File: src/UmbracoExamine.PDF/PdfTextService.cs
Comment (Shazwazza) [inline src/UmbracoExamine.PDF/PdfTextService.cs:19]: Just for backwards compat, can you obsolete this method and have a newer overload? Then you can pass `Current.Logger` into the new overload

File: src/UmbracoExamine.PDF/Properties/AssemblyInfo.cs
Comment (Shazwazza) [inline src/UmbracoExamine.PDF/Properties/AssemblyInfo.cs:1]: I think the build script probably deleted this by accident. Have seen that happen when the file is locked (i.e. when VS is open and the build script is run at the same time)

**Review by Shazwazza — CHANGES_REQUESTED** [review]
This is fantastic! I know you spent a lot of effort to write you own parsing and ended up just using another library but I assume you learned a lot :) there's just one thing to fixup if you can, if you don't have time let me know. I'll get this shipped next week!

File: src/UmbracoExamine.PDF/PdfTextService.cs
Comment (cleversolutions) [reply src/UmbracoExamine.PDF/PdfTextService.cs:30]: Oh yes, great idea. This is done.  At the time I started the text processor there wasn't a good alternative to doing it myself, now there is. I hope I can carve out some time to help with some other things.

**Comment by Shazwazza** [comment]
Yay! all merged, just cleaning some things up and we'll get a release out. Thanks so much for the effort :)

**Pull request merged**
